internal/project/golang/build.go

```diff
@@ -35,52 +47,97 @@ func NewBuild(meta *meta.Options, name, sourcePath string) *Build {
 
 // CompileDockerfile implements dockerfile.Compiler.
 func (build *Build) CompileDockerfile(output *dockerfile.Output) error {
-	stage := output.Stage(fmt.Sprintf("%s-build", build.Name())).
-		Description(fmt.Sprintf("builds %s", build.Name())).
-		From("base").
-		Step(step.Copy("/", "/").From("generate")).
-		Step(step.WorkDir(filepath.Join("/src", build.sourcePath)))
-
-	ldflags := "-s -w"
-
-	if build.meta.VersionPackage != "" {
-		stage.
-			Step(step.Arg(fmt.Sprintf("VERSION_PKG=\"%s\"", build.meta.VersionPackage))).
-			Step(step.Arg("SHA")).
-			Step(step.Arg("TAG"))
-
-		ldflags += fmt.Sprintf(" -X ${VERSION_PKG}.Name=%s", build.Name())
-		ldflags += " -X ${VERSION_PKG}.SHA=${SHA} -X ${VERSION_PKG}.Tag=${TAG}"
-	}
+	addBuildSteps := func(name string, opts CompileConfig) {
+		stage := output.Stage(fmt.Sprintf("%s-build", name)).
+			Description(fmt.Sprintf("builds %s", name)).
+			From("base").
+			Step(step.Copy("/", "/").From("generate")).
+			Step(step.WorkDir(filepath.Join("/src", build.sourcePath)))
+
+		ldflags := "-s -w"
 
-	stage.Step(step.Script(fmt.Sprintf(`go build -ldflags "%s" -o /%s`, ldflags, build.Name())).
-		MountCache(filepath.Join(build.meta.CachePath, "go-build")).
-		MountCache(filepath.Join(build.meta.GoPath, "pkg")))
+		if build.meta.VersionPackage != "" {
+			stage.
+				Step(step.Arg(fmt.Sprintf("VERSION_PKG=\"%s\"", build.meta.VersionPackage))).
+				Step(step.Arg("SHA")).
+				Step(step.Arg("TAG"))
 
-	output.Stage(build.Name()).
-		From("scratch").
-		Step(step.Copy("/"+build.Name(), "/"+build.Name()).From(fmt.Sprintf("%s-build", build.Name())))
+			ldflags += fmt.Sprintf(" -X ${VERSION_PKG}.Name=%s", build.Name())
+			ldflags += " -X ${VERSION_PKG}.SHA=${SHA} -X ${VERSION_PKG}.Tag=${TAG}"
+		}
+
+		script := step.Script(fmt.Sprintf(`go build -ldflags "%s" -o /%s`, ldflags, name)).
+			MountCache(filepath.Join(build.meta.CachePath, "go-build")).
+			MountCache(filepath.Join(build.meta.GoPath, "pkg"))
+
+		if opts != nil {
+			opts.set(script)
+		}
+
+		stage.Step(script)
+
+		output.Stage(name).
+			From("scratch").
+			Step(step.Copy("/"+name, "/"+name).From(fmt.Sprintf("%s-build", name)))
+
+		build.artifacts = append(build.artifacts, name)
+	}
+
+	if len(build.Outputs) == 0 {
+		addBuildSteps(build.Name(), nil)
+	} else {
+		for name, opts := range build.Outputs {
+			addBuildSteps(name, opts)
+		}
+	}
 
 	return nil
 }
 
 // CompileDrone implements drone.Compiler.
 func (build *Build) CompileDrone(output *drone.Output) error {
-	output.Step(drone.MakeStep(build.Name()).DependsOn(dag.GatherMatchingInputNames(build, dag.Implements((*drone.Compiler)(nil)))...))
+	for _, artifact := range build.getArtifacts() {
+		output.Step(drone.MakeStep(artifact).DependsOn(dag.GatherMatchingInputNames(build, dag.Implements((*drone.Compiler)(nil)))...))
```

Comment:
I don't want to block this PR, but it might blow up in Drone, as kube runner hates high fan-out from one step to another (stuck pipelines). It's of course Drone bug, but :shrug: 

eventually we might want to have a single step which builds them all

Reply:
I can make it depend on the unified build I guess.